packit_service/worker/handlers/github_handlers.py

```diff
@@ -235,87 +235,168 @@ def run(self) -> HandlerResults:
+            )
+            self._package_config.upstream_project_url = self.event.project_url
+        return self._package_config
 
-        return handler_results
+    @property
+    def project(self) -> GitProject:
+        if not self._project:
+            self._project = self.event.get_project()
+        return self._project
 
     def run(self) -> HandlerResults:
+        return self.copr_build_helper.run_copr_build()
+
+    def pre_check(self) -> bool:
         is_copr_build: Callable[
             [JobConfig], bool
         ] = lambda job: job.job == JobType.copr_build
 
         if self.job.job == JobType.tests and any(
             filter(is_copr_build, self.package_config.jobs)
         ):
-            return HandlerResults(
-                success=False,
-                details={
-                    "msg": "Skipping build for testing. The COPR build is defined in the config."
-                },
+            logger.info(
+                "Skipping build for testing. The COPR build is defined in the config."
             )
-        if self.event.trigger == JobTriggerType.pull_request:
-            return self.handle_pull_request()
-        # We do not support this workflow officially
-        # elif self.triggered_by == JobTriggerType.release:
-        #     self.handle_release()
-        else:
-            return HandlerResults(
-                success=False,
-                details={"msg": f"No handler for {str(self.event.trigger)}"},
+            return False
+        return True
+
+
+@add_to_mapping
+@add_to_mapping_for_job(job_type=JobType.tests)
+class ReleaseGithubCoprBuildHandler(AbstractGithubCoprBuildHandler):
+    triggers = [
+        JobTriggerType.release,
+    ]
+
+    event: ReleaseEvent
+
+    def __init__(
+        self, config: ServiceConfig, job: JobConfig, event: ReleaseEvent,
+    ):
+        super().__init__(config=config, job=job, event=event)
+        self.base_ref = event.tag_name
+
+    def pre_check(self) -> bool:
+        return (
+            super().pre_check()
+            and isinstance(self.event, ReleaseEvent)
+            and self.event.trigger == JobTriggerType.release
+        )
+
+
+@add_to_mapping
+@add_to_mapping_for_job(job_type=JobType.tests)
+class PullRequestGithubCoprBuildHandler(AbstractGithubCoprBuildHandler):
+    triggers = [
+        JobTriggerType.pull_request,
+    ]
+    event: PullRequestEvent
+
+    def __init__(
+        self, config: ServiceConfig, job: JobConfig, event: PullRequestEvent,
+    ):
+        super().__init__(config=config, job=job, event=event)
+
+    def run(self) -> HandlerResults:
+        if isinstance(self.event, GithubPullRequestHandler):
+            collaborators = self.project.who_can_merge_pr()
+            if self.event.github_login not in collaborators | self.config.admins:
+                msg = (
+                    "Only users with write or admin permissions to the repository "
+                    "can trigger Packit-as-a-Service"
```

Comment:
I've seen lots of those in Hunor's PR. Can we refactor them into a constant? for example `ADMINS_ONLY`

Reply:
Fixed in #451